### DISCUSSION TITLE
Route name on Traces and Logs

### DISCUSSION THREAD

**Discussion by alefcarlos** [dotnet/aspnet-api-versioning#1037] [Q&A] [answered]
I would like to see the version on route instead of `v{version:apiVersion}` on Traces and Logs. Is there any way ?

For example:

<img width="501" alt="image" src="https://github.com/dotnet/aspnet-api-versioning/assets/5452336/504c3773-7cd0-4d80-beaa-805554547655">

The http url does not have the selected version for the endpoint.

**Comment by xavierjohn** [comment]
Try this nuget and see if it fits your need.
https://www.nuget.org/packages/ServiceLevelIndicators.Asp.ApiVersioning

**Reply by alefcarlos** [reply]
Hey, good ?

That package adds an attribute to an specific metric, right ?

But what I mean is the display on traces/logs, it is always `v{version:apiVersion}/route-blabla`. I think it should be the version.

In my MVC controllers I always use `[Route("v{version:apiVersion}/asset-classes")]`, and it is expected to replace that pattern to the selected version.

**Reply by xavierjohn** [reply]
It emits a metrics giving the latency of each API with API version as a dimension. By attaching a TraceID, you can figure out the API version of the traces.

**Reply by alefcarlos** [reply]
I agree, but this behaviour is expected ? Should not the display be `v1/asset-classes` on traces titles ?

**Reply by xavierjohn** [reply]
That I don't know. I have only used API version as a query parameter and not in the URL. I plan to play around with Aspire soon.

**Reply by xavierjohn** [reply]
Both Aspire and VS 17. 9 is in preview so I will wait for the release version.

**Comment by commonsensesoftware** [comment]
Is there no place where the full request URL is shown in the trace? It looks like it is only showing the route template. It's not clear whether that is happening. If the endpoint was hit, then the actual request URL should be the value shown. Unless something else strange is happening in the implementation, I would expect something like `order/{id}/items` to be shown the same way. Aspire is in preview so there's always the chance this is a bug. I don't know how the values shown in the trace are generated (yet).

**Reply by alefcarlos** [reply]
It happens in other traces viewers like Jaeger, Grafana Tempo.

**Reply by alefcarlos** [reply]
In the ogs the value is parsed as expected:

```
[Scope.2]:RequestPath: /v1/weatherforecast/http-client
```

In the traces(using OpenTelemetry with console exporter) does not. The DisplayName contains the version pattern:

```
Activity.TraceId:            6e27fed42097cbc2817ea02c2f3d479a
Activity.SpanId:             5a1dc8b8796eda65
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Microsoft.AspNetCore
Activity.DisplayName:        v{version:apiVersion}/WeatherForecast/422-request
Activity.Kind:               Server
Activity.StartTime:          2023-11-21T02:50:17.7494012Z
Activity.Duration:           00:00:00.1629822
Activity.Tags:
    net.host.name: localhost
    net.host.port: 5173
    http.method: GET
    http.scheme: http
    http.target: /v1/weatherforecast/422-request
    http.url: http://localhost:5173/v1/weatherforecast/422-request
    http.flavor: 1.1
    http.user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36 Edg/119.0.0.0
    http.route: v{version:apiVersion}/WeatherForecast/422-request
    http.status_code: 422
Resource associated with Activity:
    service.name: WebApiDefault
    service.version: 1.0.0+4e98d4b0fe7a12c6448d652d1a6b431dc5eb3b7e
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.7.0-alpha.1
```

**Reply by commonsensesoftware** [reply]
This looks like it _might_ be for a _Minimal API_. Is it? It also looks like you want the value of `http.target` to show up for `Activity.DisplayName` and `http.route`. These appear to be showing you your defined route templates. I don't know there is a solution for `http.route`, but there may be a way to specify an alternate `Activity.DisplayName`. I would expect the same behavior for any route template that contains parameters; otherwise, something _special_ is happening. That doesn't make sense because the routing system has to match a request to the template. There is no in-between. When you match, you have a template against the request URL, when you don't, you only have the request URL.

**Reply by alefcarlos** [reply]
@commonsensesoftware, this example is a MVC Controller:

```chsarp
    [ApiController]
    [Route("v{version:apiVersion}/[controller]")]
    public class WeatherForecastController(ISampleApiClient sampleApiClient) : ControllerBase
    {
    }
```

Full trace:

<img width="810" alt="image" src="https://github.com/dotnet/aspnet-api-versioning/assets/5452336/862eaa03-4288-4156-a61f-7f6672a20839">

**Reply by alefcarlos** [reply]
I think this implementation is the source:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/7dff32480d8b2fe8d41e0958257c00e3602e3b3a/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L455

It uses `httpRoute`.

**Reply by commonsensesoftware** [reply, marked as answer]
I poked around a bit more and I found this line:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/7dff32480d8b2fe8d41e0958257c00e3602e3b3a/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L292

This clearly shows that it uses the `Endpoint.RoutePattern.RawText` as the value. The route template is defined as `v{version:apiVersion}/WeatherForecast` so that is what I would expect to see.  What you would probably prefer is to is `IEndpointNameMetadata`, which usually maps to `OperationId` in OpenAPI. There is also `IEndpointSummaryMetadata` and `IEndpointDescriptionMetadata`, but those probably provide longer descriptions than what you are looking for.

I don't see any hooks or other approaches to influence the trace and log values so - for now - I don't think there are any solutions. I would recommend creating an issue in the Open Telemetry repo and ask them to honor some of the other available metadata.

**Reply by alefcarlos** [reply]
I agree as well, I will open an issue there. Thank you for your attention. For now I will remove the pattern from route and explicitly set the version.

**Reply by TanvirArjel** [reply]
@alefcarlos  Did you open the issue on open-telemetry dotnet repo? Or did you find any solution?
